README.md

```diff
@@ -1,44 +1,284 @@
+- Little-endian x86_64
+- Little-endian aarch64.
+
+## Short seccomp tutorial
+
+Linux supports seccomp filters as BPF programs, that are interpreted by the
+kernel before each system call.
+
+They are installed in the kernel using 
+[`prctl(PR_SET_SECCOMP)`](https://man7.org/linux/man-pages/man2/prctl.2.html)
+or
+[`seccomp(SECCOMP_SET_MODE_FILTER)`](https://man7.org/linux/man-pages/man2/seccomp.2.html).
+
+As input, the filter receives a C struct of the following type:
+
+```C
+struct seccomp_data {
+    int nr; // syscall number.
+    __u32 arch; // arch-specific value for validation purposes.
+    __u64 instruction_pointer; // as the name suggests..
+    __u64 args[6]; // syscall arguments.
+};
+```
+
+In response, a filter returns an action, that can be either one of:
+
+```C
+#define SECCOMP_RET_KILL_PROCESS 0x80000000U /* kill the process */
+#define SECCOMP_RET_KILL_THREAD  0x00000000U /* kill the thread */
+#define SECCOMP_RET_TRAP         0x00030000U /* disallow and force a SIGSYS */
+#define SECCOMP_RET_ERRNO        0x00050000U /* returns an errno */
+#define SECCOMP_RET_TRACE        0x7ff00000U /* pass to a tracer or disallow */
+#define SECCOMP_RET_LOG          0x7ffc0000U /* allow after logging */
+#define SECCOMP_RET_ALLOW        0x7fff0000U /* allow */
+```
 
 ## Design
 
-TODO: This section should have a high-level design of the crate.
+In seccompiler, filters are defined either with a JSON file or using Rust code,
+with library-defined structures. Both representations are semantically
+equivalent and model the rules of the filter. Choosing one or the other depends
+on the use case and preference.
 
-Some questions that might help in writing this section:
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+The core of the package is the module responsible for the BPF compilation.
+It compiles seccomp filters expressed as Rust code, into BPF filters, ready to
+be loaded into the kernel. This is the seccompiler **backend.**
 
-## Usage
+The process of translating JSON filters into BPF goes through an extra step of
+deserialization and validation (the **JSON frontend**), before reaching the
+same backend for BPF codegen.
 
-TODO: This section describes how the crate is used.
+The Rust representation is therefore also an Intermediate Representation (IR)
+of the JSON filter.
+This modular implementation allows for extendability in regards to file
+formats. All that is needed is a compatible front-end.
 
-Some questions that might help in writing this section:
-- What traits do users need to implement?
-- Does the crate have any default/optional features? What is each feature
-  doing?
-- Is this crate used by other rust-vmm components? If yes, how?
+Actors 1 and 2 in the diagram below illustrate the steps required for the JSON
+and Rust filters to be compiled into BPF, respectively.
 
-## Examples
+![seccompiler architecture](./docs/img/seccompiler_arch.jpg)
```

Comment:
I think this needs to be updated to have input and output represented on the arrows. Let's chat offline.